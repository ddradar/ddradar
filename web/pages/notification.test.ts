import { mockNuxtImport } from '@nuxt/test-utils/runtime'
import Oruga from '@oruga-ui/oruga-next'
import { bulmaConfig } from '@oruga-ui/theme-bulma'
import { RouterLinkStub } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'
import { createI18n } from 'vue-i18n'

import { notifications } from '~~/../core/test/data'
import Page from '~~/pages/notification.vue'
import { mountAsync } from '~~/test/test-utils'
import { createClientPrincipal } from '~~/test/test-utils-server'

const { useEasyAuthMock, useLazyFetchMock } = vi.hoisted(() => ({
  useEasyAuthMock: vi.fn(),
  useLazyFetchMock: vi.fn(),
}))
mockNuxtImport('useEasyAuth', () => useEasyAuthMock)
mockNuxtImport('useLazyFetch', () => useLazyFetchMock)

vi.mock('~~/utils/format', async origin => {
  const actual = (await origin()) as typeof import('~~/utils/format')
  return {
    ...actual,
    unixTimeToString: (u: number) => new Date(u * 1000).toUTCString(),
  }
})

describe('Page /notification', () => {
  describe.each(['ja', 'en'])('{ locale: "%s" } snapshot test', locale => {
    const i18n = createI18n({ legacy: false, locale })

    test('({ messages: [...] }) renders notification', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useEasyAuth).mockResolvedValue({
        clientPrincipal: ref(createClientPrincipal('', '', false)),
      } as any)
      vi.mocked(useLazyFetch).mockResolvedValue({
        pending: ref(false),
        data: ref(notifications),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig], i18n],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })

    test('({ isAdmin: true }) renders Edit button', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useEasyAuth).mockResolvedValue({
        clientPrincipal: ref(createClientPrincipal('', '', true)),
      } as any)
      vi.mocked(useLazyFetch).mockResolvedValue({
        pending: ref(false),
        data: ref(notifications),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig], i18n],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.find('.mdi.mdi-pencil-box-outline').exists()).toBeTruthy()
    })

    test('({ loading: true }) renders loading state', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useEasyAuth).mockResolvedValue({
        clientPrincipal: ref(createClientPrincipal('', '', false)),
      } as any)
      vi.mocked(useLazyFetch).mockResolvedValue({
        pending: ref(true),
        data: ref(null),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig], i18n],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.find('.o-load.loading').exists()).toBeTruthy()
      expect(wrapper.find('td > section.section').element).toMatchSnapshot()
    })

    test('({ messages: [] }) renders no notification', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useEasyAuth).mockResolvedValue({
        clientPrincipal: ref(createClientPrincipal('', '', false)),
      } as any)
      vi.mocked(useLazyFetch).mockResolvedValue({
        pending: ref(false),
        data: ref([]),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig], i18n],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(
        wrapper.find('tbody > tr > td > section.section').element
      ).toMatchSnapshot()
    })
  })
})
