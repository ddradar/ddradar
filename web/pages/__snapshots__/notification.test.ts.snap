// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Page /notification > { locale: "en" } snapshot test > ({ loading: true }) renders loading state 1`] = `
<section
  class="section"
>
  <div
    class="b-skeleton is-animated"
  >
    <div
      class="b-skeleton-item is-rounded"
    />
  </div>
  <div
    class="b-skeleton is-animated"
  >
    <div
      class="b-skeleton-item is-rounded"
    />
  </div>
  <div
    class="b-skeleton is-animated"
  >
    <div
      class="b-skeleton-item is-rounded"
    />
  </div>
</section>
`;

exports[`Page /notification > { locale: "en" } snapshot test > ({ messages: [...] }) renders notification 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    All Notification List
  </h1>
  <section>
    <div
      class="b-table"
    >
      <div
        style="display: none;"
      >
        
        <span
          data-id="1"
        />
        <span
          data-id="2"
        >
          Title
        </span>
        <span
          data-id="3"
        >
          Date
        </span>
        <span
          data-id="4"
        >
          Edit
        </span>
        
      </div>
      <!--v-if-->
      <!--v-if-->
      <div
        class="table-wrapper"
      >
        <table
          class="table is-striped"
        >
          <!--v-if-->
          <thead>
            
            
            <tr>
              <th
                class="th-chevron-cell"
              />
              <!--v-if-->
              
              <th
                class=""
                draggable="false"
              >
                <span>
                   
                  <span
                    class="th-wrap sort-icon"
                    style="display: none;"
                  >
                    <span
                      class="icon is-small"
                    >
                      <i
                        class="mdi mdi-arrow-up"
                      />
                    </span>
                  </span>
                </span>
              </th>
              <th
                class=""
                draggable="false"
              >
                <span>
                  Title 
                  <span
                    class="th-wrap sort-icon"
                    style="display: none;"
                  >
                    <span
                      class="icon is-small"
                    >
                      <i
                        class="mdi mdi-arrow-up"
                      />
                    </span>
                  </span>
                </span>
              </th>
              <th
                class=""
                draggable="false"
              >
                <span>
                  Date 
                  <span
                    class="th-wrap sort-icon"
                    style="display: none;"
                  >
                    <span
                      class="icon is-small"
                    >
                      <i
                        class="mdi mdi-arrow-up"
                      />
                    </span>
                  </span>
                </span>
              </th>
              
              <!--v-if-->
            </tr>
            <!--v-if-->
            <!--v-if-->
          </thead>
          <tbody>
            
            
            <tr
              class=""
              draggable="false"
            >
              <td
                class="chevron-cell"
              >
                <span
                  class="icon is-clickable"
                  role="button"
                >
                  <i
                    class="mdi mdi-chevron-right mdi-24px"
                  />
                </span>
              </td>
              <!--v-if-->
              
              <td
                class=""
              >
                <button
                  class="button is-info"
                  type="button"
                >
                  <span
                    class="button-wrapper"
                  >
                    <span
                      class="icon"
                    >
                      <i
                        class="mdi mdi-info mdi-24px"
                      />
                    </span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </button>
              </td>
              <td
                class=""
                data-label="Title"
              >
                新曲を追加しました
              </td>
              <td
                class=""
                data-label="Date"
              >
                Thu, 13 Aug 2020 00:00:00 GMT
              </td>
              
              <!--v-if-->
            </tr>
            <transition-stub
              appear="false"
              css="true"
              name=""
              persisted="false"
            >
              <!--v-if-->
            </transition-stub>
            <!--v-if-->
            
            
            <tr
              class=""
              draggable="false"
            >
              <td
                class="chevron-cell"
              >
                <span
                  class="icon is-clickable"
                  role="button"
                >
                  <i
                    class="mdi mdi-chevron-right mdi-24px"
                  />
                </span>
              </td>
              <!--v-if-->
              
              <td
                class=""
              >
                <button
                  class="button is-warning"
                  type="button"
                >
                  <span
                    class="button-wrapper"
                  >
                    <span
                      class="icon"
                    >
                      <i
                        class="mdi mdi-warning mdi-24px"
                      />
                    </span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </button>
              </td>
              <td
                class=""
                data-label="Title"
              >
                このサイトはベータ版です
              </td>
              <td
                class=""
                data-label="Date"
              >
                Sat, 01 Aug 2020 03:00:00 GMT
              </td>
              
              <!--v-if-->
            </tr>
            <transition-stub
              appear="false"
              css="true"
              name=""
              persisted="false"
            >
              <!--v-if-->
            </transition-stub>
            <!--v-if-->
            
            
            <tr
              class=""
              draggable="false"
            >
              <td
                class="chevron-cell"
              >
                <span
                  class="icon is-clickable"
                  role="button"
                >
                  <i
                    class="mdi mdi-chevron-right mdi-24px"
                  />
                </span>
              </td>
              <!--v-if-->
              
              <td
                class=""
              >
                <button
                  class="button is-info"
                  type="button"
                >
                  <span
                    class="button-wrapper"
                  >
                    <span
                      class="icon"
                    >
                      <i
                        class="mdi mdi-info mdi-24px"
                      />
                    </span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </button>
              </td>
              <td
                class=""
                data-label="Title"
              >
                v0.6.0をリリースしました
              </td>
              <td
                class=""
                data-label="Date"
              >
                Tue, 11 Aug 2020 03:00:00 GMT
              </td>
              
              <!--v-if-->
            </tr>
            <transition-stub
              appear="false"
              css="true"
              name=""
              persisted="false"
            >
              <!--v-if-->
            </transition-stub>
            <!--v-if-->
            
            
            <!--v-if-->
          </tbody>
          <!--v-if-->
        </table>
        
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <!--v-if-->
        </transition-stub>
        
      </div>
      
      <div
        class="level"
      >
        <div>
          
          
        </div>
        <div>
          <div>
            <nav
              class="pagination is-right"
            >
              <a
                aria-current="false"
                class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
                href="#"
                role="button"
              >
                
                <span
                  aria-hidden="true"
                  class="icon"
                >
                  <i
                    class="mdi mdi-chevron-left mdi-24px"
                  />
                </span>
                
              </a>
              <a
                aria-current="false"
                class="pagination-link pagination-next is-disabled pagination-next is-disabled"
                href="#"
                role="button"
              >
                
                <span
                  aria-hidden="true"
                  class="icon"
                >
                  <i
                    class="mdi mdi-chevron-right mdi-24px"
                  />
                </span>
                
              </a>
              <ul
                class="pagination-list"
              >
                <!--First-->
                <!--v-if-->
                <!--v-if-->
                <!--Pages-->
                
                <li>
                  <a
                    aria-current="true"
                    class="pagination-link is-current"
                    href="#"
                    role="button"
                  >
                    
                    1
                    
                  </a>
                </li>
                
                <!--Last-->
                <!--v-if-->
                <!--v-if-->
              </ul>
            </nav>
          </div>
        </div>
      </div>
      
    </div>
  </section>
</section>
`;

exports[`Page /notification > { locale: "en" } snapshot test > ({ messages: [] }) renders no notification 1`] = `
<section
  class="section"
>
  <div
    class="content has-text-grey has-text-centered"
  >
    <p>
      No Data
    </p>
  </div>
</section>
`;

exports[`Page /notification > { locale: "ja" } snapshot test > ({ loading: true }) renders loading state 1`] = `
<section
  class="section"
>
  <div
    class="b-skeleton is-animated"
  >
    <div
      class="b-skeleton-item is-rounded"
    />
  </div>
  <div
    class="b-skeleton is-animated"
  >
    <div
      class="b-skeleton-item is-rounded"
    />
  </div>
  <div
    class="b-skeleton is-animated"
  >
    <div
      class="b-skeleton-item is-rounded"
    />
  </div>
</section>
`;

exports[`Page /notification > { locale: "ja" } snapshot test > ({ messages: [...] }) renders notification 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    お知らせ一覧
  </h1>
  <section>
    <div
      class="b-table"
    >
      <div
        style="display: none;"
      >
        
        <span
          data-id="1"
        />
        <span
          data-id="2"
        >
          タイトル
        </span>
        <span
          data-id="3"
        >
          投稿日時
        </span>
        <span
          data-id="4"
        >
          Edit
        </span>
        
      </div>
      <!--v-if-->
      <!--v-if-->
      <div
        class="table-wrapper"
      >
        <table
          class="table is-striped"
        >
          <!--v-if-->
          <thead>
            
            
            <tr>
              <th
                class="th-chevron-cell"
              />
              <!--v-if-->
              
              <th
                class=""
                draggable="false"
              >
                <span>
                   
                  <span
                    class="th-wrap sort-icon"
                    style="display: none;"
                  >
                    <span
                      class="icon is-small"
                    >
                      <i
                        class="mdi mdi-arrow-up"
                      />
                    </span>
                  </span>
                </span>
              </th>
              <th
                class=""
                draggable="false"
              >
                <span>
                  タイトル 
                  <span
                    class="th-wrap sort-icon"
                    style="display: none;"
                  >
                    <span
                      class="icon is-small"
                    >
                      <i
                        class="mdi mdi-arrow-up"
                      />
                    </span>
                  </span>
                </span>
              </th>
              <th
                class=""
                draggable="false"
              >
                <span>
                  投稿日時 
                  <span
                    class="th-wrap sort-icon"
                    style="display: none;"
                  >
                    <span
                      class="icon is-small"
                    >
                      <i
                        class="mdi mdi-arrow-up"
                      />
                    </span>
                  </span>
                </span>
              </th>
              
              <!--v-if-->
            </tr>
            <!--v-if-->
            <!--v-if-->
          </thead>
          <tbody>
            
            
            <tr
              class=""
              draggable="false"
            >
              <td
                class="chevron-cell"
              >
                <span
                  class="icon is-clickable"
                  role="button"
                >
                  <i
                    class="mdi mdi-chevron-right mdi-24px"
                  />
                </span>
              </td>
              <!--v-if-->
              
              <td
                class=""
              >
                <button
                  class="button is-info"
                  type="button"
                >
                  <span
                    class="button-wrapper"
                  >
                    <span
                      class="icon"
                    >
                      <i
                        class="mdi mdi-info mdi-24px"
                      />
                    </span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </button>
              </td>
              <td
                class=""
                data-label="タイトル"
              >
                新曲を追加しました
              </td>
              <td
                class=""
                data-label="投稿日時"
              >
                Thu, 13 Aug 2020 00:00:00 GMT
              </td>
              
              <!--v-if-->
            </tr>
            <transition-stub
              appear="false"
              css="true"
              name=""
              persisted="false"
            >
              <!--v-if-->
            </transition-stub>
            <!--v-if-->
            
            
            <tr
              class=""
              draggable="false"
            >
              <td
                class="chevron-cell"
              >
                <span
                  class="icon is-clickable"
                  role="button"
                >
                  <i
                    class="mdi mdi-chevron-right mdi-24px"
                  />
                </span>
              </td>
              <!--v-if-->
              
              <td
                class=""
              >
                <button
                  class="button is-warning"
                  type="button"
                >
                  <span
                    class="button-wrapper"
                  >
                    <span
                      class="icon"
                    >
                      <i
                        class="mdi mdi-warning mdi-24px"
                      />
                    </span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </button>
              </td>
              <td
                class=""
                data-label="タイトル"
              >
                このサイトはベータ版です
              </td>
              <td
                class=""
                data-label="投稿日時"
              >
                Sat, 01 Aug 2020 03:00:00 GMT
              </td>
              
              <!--v-if-->
            </tr>
            <transition-stub
              appear="false"
              css="true"
              name=""
              persisted="false"
            >
              <!--v-if-->
            </transition-stub>
            <!--v-if-->
            
            
            <tr
              class=""
              draggable="false"
            >
              <td
                class="chevron-cell"
              >
                <span
                  class="icon is-clickable"
                  role="button"
                >
                  <i
                    class="mdi mdi-chevron-right mdi-24px"
                  />
                </span>
              </td>
              <!--v-if-->
              
              <td
                class=""
              >
                <button
                  class="button is-info"
                  type="button"
                >
                  <span
                    class="button-wrapper"
                  >
                    <span
                      class="icon"
                    >
                      <i
                        class="mdi mdi-info mdi-24px"
                      />
                    </span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </button>
              </td>
              <td
                class=""
                data-label="タイトル"
              >
                v0.6.0をリリースしました
              </td>
              <td
                class=""
                data-label="投稿日時"
              >
                Tue, 11 Aug 2020 03:00:00 GMT
              </td>
              
              <!--v-if-->
            </tr>
            <transition-stub
              appear="false"
              css="true"
              name=""
              persisted="false"
            >
              <!--v-if-->
            </transition-stub>
            <!--v-if-->
            
            
            <!--v-if-->
          </tbody>
          <!--v-if-->
        </table>
        
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <!--v-if-->
        </transition-stub>
        
      </div>
      
      <div
        class="level"
      >
        <div>
          
          
        </div>
        <div>
          <div>
            <nav
              class="pagination is-right"
            >
              <a
                aria-current="false"
                class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
                href="#"
                role="button"
              >
                
                <span
                  aria-hidden="true"
                  class="icon"
                >
                  <i
                    class="mdi mdi-chevron-left mdi-24px"
                  />
                </span>
                
              </a>
              <a
                aria-current="false"
                class="pagination-link pagination-next is-disabled pagination-next is-disabled"
                href="#"
                role="button"
              >
                
                <span
                  aria-hidden="true"
                  class="icon"
                >
                  <i
                    class="mdi mdi-chevron-right mdi-24px"
                  />
                </span>
                
              </a>
              <ul
                class="pagination-list"
              >
                <!--First-->
                <!--v-if-->
                <!--v-if-->
                <!--Pages-->
                
                <li>
                  <a
                    aria-current="true"
                    class="pagination-link is-current"
                    href="#"
                    role="button"
                  >
                    
                    1
                    
                  </a>
                </li>
                
                <!--Last-->
                <!--v-if-->
                <!--v-if-->
              </ul>
            </nav>
          </div>
        </div>
      </div>
      
    </div>
  </section>
</section>
`;

exports[`Page /notification > { locale: "ja" } snapshot test > ({ messages: [] }) renders no notification 1`] = `
<section
  class="section"
>
  <div
    class="content has-text-grey has-text-centered"
  >
    <p>
      データがありません
    </p>
  </div>
</section>
`;
